### PR TITLE
feat-SER-67-restrict-invalid-domains

### PR DESCRIPTION
**Pull request by HarshMN2345**
## Test Plan

This PR introduces a **custom domain validator** for `.appwrite.network` domains to enforce stricter rules and improve security and consistency.

### ✅ Key Changes

- **New Validator**: `AppwriteDomain`
  - Allows only **single-level subdomains** (e.g., `app.appwrite.network`)
  - Rejects **sub-subdomains** (e.g., `dev.test.appwrite.network`)
  - Rejects **domains with spaces** or other invalid formats

    (read dynamically via `System::getEnv('_APP_DOMAIN_SITES', 'appwrite.network')`)

- **Applied In**:
  - `src/Appwrite/Platform/Modules/Console/Http/Resources/Get.php`
  - `src/Appwrite/Platform/Modules/Proxy/Http/Rules/API/Create.php`
  - `src/Appwrite/Platform/Modules/Proxy/Http/Rules/Function/Create.php`
  - `src/Appwrite/Platform/Modules/Proxy/Http/Rules/Redirect/Create.php`
  - `src/Appwrite/Platform/Modules/Proxy/Http/Rules/Site/Create.php`

- **Unit Tests**:
  - Added thorough PHPUnit tests for:
    - Valid and invalid `.appwrite.network` domains
    - Domains with multiple subdomains
    - Domains containing whitespace
    - Non-Appwrite domains
    - Edge cases (`null`, empty string, invalid types)

## Related PRs and Issues

- (Related PR or issue)

## Checklist

- [x] Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?
- [x] If the PR includes a change to an API's metadata (desc, label, params, etc.), does it also include updated API specs and example docs?

### CODE REVIEW

File: src/Appwrite/Platform/Modules/Proxy/Http/Rules/API/Create.php [hunk 1]
```diff
@@ -71,6 +72,19 @@ public function __construct()
...
+        try {
+            $domain = new Domain($domain);
+        } catch (\Throwable) {
+            throw new Exception(Exception::GENERAL_ARGUMENT_INVALID, 'Domain may not start with http:// or https://.');
+        }
+
+        // 2. Domain prefix validations
+        if (\str_starts_with($domain->get(), 'commit-') || \str_starts_with($domain->get(), 'branch-')) {
+            throw new Exception(Exception::GENERAL_ARGUMENT_INVALID, 'This domain name is not allowed. Please use a different domain.');
+        }
```
Comment (abnegate) [inline src/Appwrite/Platform/Modules/Proxy/Http/Rules/API/Create.php]: This check should be part of one of the validators

File: src/Appwrite/Platform/Modules/Proxy/Http/Rules/API/Create.php [hunk 2]
```diff
@@ -71,6 +72,19 @@ public function __construct()
 
     public function action(string $domain, Response $response, Document $project, Certificate $queueForCertificates, Event $queueForEvents, Database $dbForPlatform)
     {
+        // 1. Domain format validations
+        try {
+            $domain = new Domain($domain);
+        } catch (\Throwable) {
+            throw new Exception(Exception::GENERAL_ARGUMENT_INVALID, 'Domain may not start with http:// or https://.');
```
Comment (abnegate) [inline src/Appwrite/Platform/Modules/Proxy/Http/Rules/API/Create.php]: I don't think we can assume that the issue was that it included http or https, we should use the error message

File: src/Appwrite/Platform/Modules/Proxy/Http/Rules/API/Create.php [hunk 3]
```diff
@@ -60,7 +61,7 @@ public function __construct()
             ->label('abuse-limit', 10)
             ->label('abuse-key', 'userId:{userId}, url:{url}')
             ->label('abuse-time', 60)
-            ->param('domain', null, new ValidatorDomain(), 'Domain name.')
+            ->param('domain', null, new AnyOf([new ValidatorDomain(), new AppwriteDomain()]), 'Domain name.')
```
Comment (abnegate) [inline src/Appwrite/Platform/Modules/Proxy/Http/Rules/API/Create.php]: Should this be AllOf? This will succeed even if `AppwriteDomain` validates as false

File: src/Appwrite/Platform/Modules/Console/Http/Resources/Get.php [hunk 4]
```diff
@@ -67,10 +68,11 @@ public function action(
         Database $dbForPlatform
     ) {
         if ($type === 'rules') {
-            $validator = new Domain($value);
+            $domainValidator = new Domain($value);
+            $appwriteDomainValidator = new AppwriteDomain();
```
Comment (abnegate) [inline src/Appwrite/Platform/Modules/Console/Http/Resources/Get.php]: It looks like we always use both, can AppwriteDomain extend Domain?

File: app/init/constants.php [hunk 5]
```diff
@@ -4,6 +4,7 @@
 
 const APP_NAME = 'Appwrite';
 const APP_DOMAIN = 'appwrite.io';
+const APP_DOMAIN_SITES_SUFFIX = 'appwrite.network'; // Default suffix for Appwrite sites domains
```
Comment (vermakhushboo) [inline app/init/constants.php]: We already have `_APP_DOMAIN_SITES` for this.

File: src/Appwrite/Network/Validator/AppwriteDomain.php [hunk 6]
```diff
@@ -0,0 +1,152 @@
+<?php
+
+namespace Appwrite\Network\Validator;
```
Comment (vermakhushboo) [inline src/Appwrite/Network/Validator/AppwriteDomain.php]: This file should probably be in Appwrite\Domain\Validator path

File: src/Appwrite/Network/Validator/AppwriteDomain.php [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+use Utopia\System\System;
+use Utopia\Validator\Domain as ValidatorDomain;
+
+class AppwriteDomain extends ValidatorDomain
+{
+    public function getDescription(): string
+    {
+        $sitesDomain = System::getEnv('_APP_DOMAIN_SITES', defined('APP_DOMAIN_SITES_SUFFIX') ? APP_DOMAIN_SITES_SUFFIX : 'appwrite.network');
```
Comment (vermakhushboo) [inline src/Appwrite/Network/Validator/AppwriteDomain.php]: Again, we should use _APP_DOMAIN_SITES for it. Locally, it is `sites.localhost`

File: src/Appwrite/Network/Validator/AppwriteDomain.php [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+        return "Value must be a valid domain name. For Appwrite-managed domains (e.g., {$sitesDomain}, functions.localhost, appwrite.network), only one-level subdomain is allowed.";
```
Comment (vermakhushboo) [inline src/Appwrite/Network/Validator/AppwriteDomain.php]: Don't hard-code any values. It should come from env var.

File: src/Appwrite/Network/Validator/AppwriteDomain.php [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+        // 10. Global forbidden prefixes check (commit-, branch-)
+        $firstLabel = $parts[0];
+        $firstLabelLower = strtolower($firstLabel);
+        if (str_starts_with($firstLabelLower, 'commit-') || str_starts_with($firstLabelLower, 'branch-')) {
+            return false;
+        }
+
+        // 11. Appwrite-managed domains: only single-level subdomains, no forbidden prefixes
+        $functionsDomain = System::getEnv('_APP_DOMAIN_FUNCTIONS', 'functions.localhost');
+        $sitesDomain = System::getEnv('_APP_DOMAIN_SITES', defined('APP_DOMAIN_SITES_SUFFIX') ? APP_DOMAIN_SITES_SUFFIX : 'appwrite.network');
```
Comment (vermakhushboo) [inline src/Appwrite/Network/Validator/AppwriteDomain.php]: Only use `_APP_DOMAIN_SITES` here as well

File: src/Appwrite/Network/Validator/AppwriteDomain.php [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+        $appwriteDomain = defined('APP_DOMAIN_SITES_SUFFIX') ? APP_DOMAIN_SITES_SUFFIX : 'appwrite.network';
```
Comment (vermakhushboo) [inline src/Appwrite/Network/Validator/AppwriteDomain.php]: Don't we have `_APP_DOMAIN` for `_APP_CONSOLE_DOMAIN` for this?

File: src/Appwrite/Network/Validator/AppwriteDomain.php [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+
+        // Check for specific localhost domains that should be rejected
+        if ($domain === 'localhost' || $domain === 'api.localhost') {
```
Comment (vermakhushboo) [inline src/Appwrite/Network/Validator/AppwriteDomain.php]: I don't think we need this check

File: src/Appwrite/Network/Validator/AppwriteDomain.php [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+                }
+            }
+        }
+
+        // 12. RFC domain validation (parent)
+        if (!parent::isValid($value)) {
+            return false;
+        }
+
+        // 13. All other domains: already validated above
```
Comment (vermakhushboo) [inline src/Appwrite/Network/Validator/AppwriteDomain.php]: We don't need these comments where code is self-explanatory

File: src/Appwrite/Platform/Modules/Console/Http/Resources/Get.php [hunk 13]
```diff
@@ -67,10 +68,11 @@ public function action(
         Database $dbForPlatform
     ) {
         if ($type === 'rules') {
-            $validator = new Domain($value);
+            $domainValidator = new Domain($value);
```
Comment (vermakhushboo) [inline src/Appwrite/Platform/Modules/Console/Http/Resources/Get.php]: When you are extending the `Domain` class, then you don't need to use both. Just use `AppwriteDomainValidator`

File: src/Appwrite/Platform/Modules/Proxy/Http/Rules/API/Create.php [hunk 14]
```diff
@@ -71,6 +71,14 @@ public function __construct()
 
     public function action(string $domain, Response $response, Document $project, Certificate $queueForCertificates, Event $queueForEvents, Database $dbForPlatform)
     {
+        // 1. Domain format validations
+        try {
+            $domain = new Domain($domain);
```
Comment (vermakhushboo) [inline src/Appwrite/Platform/Modules/Proxy/Http/Rules/API/Create.php]: You don't need to do this. It already happens in the param validation.

File: src/Appwrite/Platform/Modules/Proxy/Http/Rules/Function/Create.php [hunk 15]
```diff
@@ -76,6 +76,14 @@ public function __construct()
 
     public function action(string $domain, string $functionId, string $branch, Response $response, Document $project, Certificate $queueForCertificates, Event $queueForEvents, Database $dbForPlatform, Database $dbForProject)
     {
+        // 1. Domain format validations
+        try {
```
Comment (vermakhushboo) [inline src/Appwrite/Platform/Modules/Proxy/Http/Rules/Function/Create.php]: Same comment as above

File: src/Appwrite/Platform/Modules/Proxy/Http/Rules/Redirect/Create.php [hunk 16]
```diff
@@ -79,6 +79,14 @@ public function __construct()
 
     public function action(string $domain, string $url, int $statusCode, string $resourceId, string $resourceType, Response $response, Document $project, Certificate $queueForCertificates, Event $queueForEvents, Database $dbForPlatform, Database $dbForProject)
     {
+        // 1. Domain format validations
+        try {
```
Comment (vermakhushboo) [inline src/Appwrite/Platform/Modules/Proxy/Http/Rules/Redirect/Create.php]: Here as well

File: tests/unit/Network/Validators/AppwriteDomainTest.php [hunk 17]
```diff
@@ -0,0 +1,188 @@
+<?php
+
+namespace Tests\Unit\Network\Validators;
+
+use Appwrite\Network\Validator\AppwriteDomain;
```
Comment (vermakhushboo) [inline tests/unit/Network/Validators/AppwriteDomainTest.php]: As mentioned above, this should not be under `Network`. It should be under `Domain`.

File: tests/unit/Network/Validators/AppwriteDomainTest.php [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+    }
+
+    public function tearDown(): void
+    {
+        $this->validator = null;
+    }
+
+    public function testIsValid(): void
+    {
+        // Get the actual configured suffix from environment (in Docker it's sites.localhost)
```
Comment (vermakhushboo) [inline tests/unit/Network/Validators/AppwriteDomainTest.php]: Please clean up the comments. Only add where necessary.

File: src/Appwrite/Domain/Validator/AppwriteDomain.php [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+            return false;
+        }
+
+        $managedDomains = [
+            System::getEnv('_APP_DOMAIN_FUNCTIONS'),
+            System::getEnv('_APP_DOMAIN_SITES')
+        ];
+
+        foreach ($managedDomains as $managedDomain) {
+            if (empty($managedDomain)) {
```
Comment (vermakhushboo) [inline src/Appwrite/Domain/Validator/AppwriteDomain.php:52]: Not needed

File: src/Appwrite/Platform/Modules/Console/Http/Resources/Get.php [hunk 20]
```diff
@@ -67,10 +67,10 @@ public function action(
         Database $dbForPlatform
     ) {
         if ($type === 'rules') {
-            $validator = new Domain($value);
+            $appwriteDomainValidator = new AppwriteDomain();
 
-            if (!$validator->isValid($value)) {
-                throw new Exception(Exception::GENERAL_ARGUMENT_INVALID, $validator->getDescription());
+            if (!$appwriteDomainValidator->isValid($value)) {
+                throw new Exception(Exception::GENERAL_ARGUMENT_INVALID, 'Value must be a valid domain name or a valid Appwrite subdomain.');
```
Comment (vermakhushboo) [inline src/Appwrite/Platform/Modules/Console/Http/Resources/Get.php:73]: This should come from validator.

File: src/Appwrite/Platform/Modules/Proxy/Http/Rules/API/Create.php [hunk 21]
```diff
@@ -71,8 +71,11 @@ public function __construct()
 
     public function action(string $domain, Response $response, Document $project, Certificate $queueForCertificates, Event $queueForEvents, Database $dbForPlatform)
     {
+
+        $domainString = strtolower($domain);
```
Comment (vermakhushboo) [inline src/Appwrite/Platform/Modules/Proxy/Http/Rules/API/Create.php]: This is redundant as this is already happening inside validator

File: src/Appwrite/Platform/Modules/Proxy/Http/Rules/API/Create.php [hunk 22]
```diff
@@ -71,8 +71,11 @@ public function __construct()
...
+
+
         $deniedDomains = [
-            'localhost',
```
Comment (vermakhushboo) [inline src/Appwrite/Platform/Modules/Proxy/Http/Rules/API/Create.php]: If it was denied here, then don't modify this logic.

File: src/Appwrite/Platform/Modules/Proxy/Http/Rules/API/Create.php [hunk 23]
```diff
@@ -98,25 +101,15 @@ public function action(string $domain, Response $response, Document $project, Ce
             $deniedDomains[] = $denyListDomain;
         }
 
-        if (\in_array($domain, $deniedDomains)) {
+        if (\in_array($domainString, $deniedDomains)) {
```
Comment (vermakhushboo) [inline src/Appwrite/Platform/Modules/Proxy/Http/Rules/API/Create.php]: No need to rename variable

File: src/Appwrite/Platform/Modules/Proxy/Http/Rules/API/Create.php [hunk 24]
```diff
@@ -98,25 +101,15 @@ public function action(string $domain, Response $response, Document $project, Ce
...
             throw new Exception(Exception::GENERAL_ARGUMENT_INVALID, 'This domain name is not allowed. Please use a different domain.');
         }
 
-        if (\str_starts_with($domain, 'commit-') || \str_starts_with($domain, 'branch-')) {
-            throw new Exception(Exception::GENERAL_ARGUMENT_INVALID, 'This domain name is not allowed. Please use a different domain.');
-        }
-
-        try {
-            $domain = new Domain($domain);
```
Comment (vermakhushboo) [inline src/Appwrite/Platform/Modules/Proxy/Http/Rules/API/Create.php:110]: Don't remove this code. Replace `Domain` with `AppwriteDomain`

File: src/Appwrite/Platform/Modules/Proxy/Http/Rules/API/Create.php [hunk 25]
```diff
@@ -98,25 +101,15 @@ public function action(string $domain, Response $response, Document $project, Ce
...
-        } catch (\Throwable) {
-            throw new Exception(Exception::GENERAL_ARGUMENT_INVALID, 'Domain may not start with http:// or https://.');
-        }
-
         // TODO: @christyjacob remove once we migrate the rules in 1.7.x
-        $ruleId = System::getEnv('_APP_RULES_FORMAT') === 'md5' ? md5($domain->get()) : ID::unique();
+        $ruleId = System::getEnv('_APP_RULES_FORMAT') === 'md5' ? md5($domainString) : ID::unique();
```
Comment (vermakhushboo) [inline src/Appwrite/Platform/Modules/Proxy/Http/Rules/API/Create.php]: This should not be touched.

File: src/Appwrite/Platform/Modules/Proxy/Http/Rules/Site/Create.php [hunk 26]
```diff
@@ -177,7 +170,7 @@ public function action(string $domain, string $siteId, string $branch, Response
             'deploymentResourceInternalId' => $site->getSequence(),
             'deploymentVcsProviderBranch' => $branch,
             'certificateId' => '',
-            'search' => implode(' ', [$ruleId, $domain->get(), $branch]),
+            'search' => implode(' ', [$ruleId, $domainString, $branch]),
```
Comment (vermakhushboo) [inline src/Appwrite/Platform/Modules/Proxy/Http/Rules/Site/Create.php]: I don't think we should need all these changes

File: tests/unit/Network/Validators/AppwriteDomainTest.php [hunk 27]
```diff
@@ -0,0 +1,146 @@
...
+            $this->assertEquals(false, $this->validator->isValid('BRANCH-test.' . $sitesDomain));
+
+            $this->assertEquals(true, $this->validator->isValid('commitment.' . $sitesDomain));
+            $this->assertEquals(true, $this->validator->isValid('branching.' . $sitesDomain));
+            $this->assertEquals(true, $this->validator->isValid('my-commit.' . $sitesDomain));
+            $this->assertEquals(true, $this->validator->isValid('my-branch.' . $sitesDomain));
+            $this->assertEquals(true, $this->validator->isValid('pre-commit.' . $sitesDomain));
+            $this->assertEquals(true, $this->validator->isValid('post-branch.' . $sitesDomain));
+
+            $this->assertEquals(false, $this->validator->isValid('.api.' . $sitesDomain));
```
Comment (vermakhushboo) [inline tests/unit/Network/Validators/AppwriteDomainTest.php]: This is redundant

File: src/Appwrite/Platform/Modules/Console/Http/Resources/Get.php (hunk 20 above)
Comment (Meldiron) [reply src/Appwrite/Platform/Modules/Console/Http/Resources/Get.php:73]: We can use `$appwriteDomainValidator->getMessage` here

File: tests/unit/Domain/Validators/AppwriteDomainTest.php [hunk 28]
```diff
@@ -0,0 +1,146 @@
...
+    }
+
+    public function testIsValid(): void
+    {
+        $sitesDomain = \Utopia\System\System::getEnv('_APP_DOMAIN_SITES');
+        $functionsDomain = \Utopia\System\System::getEnv('_APP_DOMAIN_FUNCTIONS');
+
+        if (!empty($sitesDomain)) {
+            $this->assertEquals(true, $this->validator->isValid('api.' . $sitesDomain));
+            $this->assertEquals(true, $this->validator->isValid('test.' . $sitesDomain));
```
Comment (Meldiron) [inline tests/unit/Domain/Validators/AppwriteDomainTest.php:29]: Those are unit tests, nice!
Let's also add e2e tests, in Proxy tests to try to create domains with sub-sub domains, expecting 400 status code

File: src/Appwrite/Domain/Validator/AppwriteDomain.php [hunk 29]
```diff
@@ -0,0 +1,81 @@
+<?php
+
+namespace Appwrite\Domain\Validator;
+
+use Utopia\System\System;
+use Utopia\Validator\Domain;
+
+class AppwriteDomain extends Domain
```
Comment (Meldiron) [inline src/Appwrite/Domain/Validator/AppwriteDomain.php:8]: This validator sounds too complex.

Some things sound related to domains in general, not our subdomain issue. For that, inside isValid, we could run parent::isValid(), to run standard domain's validation.

What I would expect in this new validator is just a constructor with integer of amount of domains allowed (top-level domain, domain, subdomain, sub-sub domain...). And an isValid that splits by dot and compares count.
Due to how simple this it, we might want to add this to original Domain validator, without need to make a new one - just adding 1 if check on top of what's already there, and to make amount of subdomians configurable, add parameter to constructor. From Appwrite we can then set `domains: 3`, to only allow to have 3 parts like `myapp.appwrite.network`.

**Comment by Meldiron** [comment]
After looong discussions, we now settled on an approach. Closing in favour of: https://github.com/appwrite/appwrite/pull/10374
